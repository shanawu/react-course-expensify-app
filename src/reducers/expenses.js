// Expense Reducer

const expensesReducerDefaultState = [];

export default (state = expensesReducerDefaultState, action) => {
  switch (action.type) {
    // add expense
    case "ADD_EXPENSE":
      return [...state, action.expense]; // spread operator to add the newly added expense without modifying the original value
    // remove expense
    case "REMOVE_EXPENSE":
      return state.filter(({ id }) => id !== action.id); // state.filter((expense) => expense.id !== action.expense.id)
    // edit expense
    case "EDIT_EXPENSE":
      return state.map((expense) => {
        if (expense.id === action.id) {
          // finding if there is a match in the expenses
          return {
            ...expense, // spread the expense
            ...action.updates, // spread the updates
          };
        } else {
          return expense;
        }
      });
    default:
      return state;
  }
};
