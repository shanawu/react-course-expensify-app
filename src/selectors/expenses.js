import moment from "moment";
// Get Filtered Expenses

export default (expenses, { text, sortBy, startDate, endDate }) => {
  return expenses
    .filter((expense) => {
      const createdAtMoment = moment(expense.createdAt);
      const startDateMatch = startDate
        ? startDate.isSameOrBefore(createdAtMoment, "day")
        : true;
      const endDateMatch = endDate
        ? endDate.isSameOrAfter(createdAtMoment, "day")
        : true;
      const textMatch =
        typeof text === undefined ||
        expense.description.toLowerCase().indexOf(text.toLowerCase()) >= 0;
      // typeof text === undefined || expense.description.toLowerCase().includes(text.toLowerCase())

      return startDateMatch && endDateMatch && textMatch;
    })
    .sort((a, b) => {
      if (sortBy === "date") {
        return a.createdAt < b.createdAt ? 1 : -1; // 1: b come first, -1: a come first
      } else if (sortBy === "amount") {
        return a.amount < b.amount ? 1 : -1;
      }
    });
};
