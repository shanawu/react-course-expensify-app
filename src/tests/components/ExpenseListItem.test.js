import React from "react";
import { render } from "@testing-library/react";
import { BrowserRouter } from "react-router-dom";
import ExpenseListItem from "../../components/ExpenseListItem";
import expenses from "../fixtures/expenses";

test("should render ExpenseListItem with fixture data", () => {
  const { container } = render(
    <BrowserRouter basename="/">
      <ExpenseListItem {...expenses[0]} />
    </BrowserRouter>
  );
  expect(container).toMatchSnapshot();
});
