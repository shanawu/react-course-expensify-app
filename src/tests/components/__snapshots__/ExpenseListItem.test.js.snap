// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render ExpenseListItem with fixture data 1`] = `
<div>
  <div>
    <a
      href="/edit/1"
    >
      Gas
    </a>
    <p>
      156
       - 
      0
    </p>
  </div>
</div>
`;
